# Library

In [3]:
import pandas as pd
import json
import urllib.request, json 
import requests
from datetime import datetime
import numpy as np

In [10]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


# Province Detail Dictionary

Province Key

In [4]:
province_code = {0: 'ACEH', 1: 'BALI', 2: 'BANTEN', 3: 'BENGKULU', 4: 'DAERAH_ISTIMEWA_YOGYAKARTA', 5: 'DKI_JAKARTA', 
                      6: 'GORONTALO', 7: 'JAMBI', 8: 'JAWA_BARAT', 9: 'JAWA_TENGAH', 10: 'JAWA_TIMUR', 11: 'KALIMANTAN_BARAT', 
                      12: 'KALIMANTAN_SELATAN', 13: 'KALIMANTAN_TENGAH', 14: 'KALIMANTAN_TIMUR', 15: 'KALIMANTAN_UTARA', 
                      16: 'KEPULAUAN_BANGKA_BELITUNG', 17: 'KEPULAUAN_RIAU', 18: 'LAMPUNG', 19: 'MALUKU', 20: 'MALUKU_UTARA', 
                      21: 'NUSA_TENGGARA_BARAT', 22: 'NUSA_TENGGARA_TIMUR', 23: 'PAPUA', 24: 'PAPUA_BARAT', 25: 'RIAU', 
                      26: 'SULAWESI_BARAT', 27: 'SULAWESI_SELATAN', 28: 'SULAWESI_TENGAH', 29: 'SULAWESI_TENGGARA', 
                      30: 'SULAWESI_UTARA', 31: 'SUMATERA_BARAT', 32: 'SUMATERA_SELATAN', 33 : 'SUMATERA_UTARA'}

Polygon Border Point

In [5]:
polygon_border = {
 'north': [6.2744496,-8.0615889, -5.4996381, -2.2771362, -7.5412887, -4.9993635, 1.3647141, -0.5088816,
  -4.0387936, -4.0387936, -4.8926893, 2.3146446, -1.3024076, 0.7954077, 2.569023, 4.407823, -0.278773, 4.9918231, -3.7237393,
  -2.2664176, 3.4464276, -6.7458825, -6.7574264, 1.1369057, 1.2874153, 3.0028667, -0.22745, -1.8906412, 1.583554, -2.7702794,
  5.7639518, 0.9073892, -1.5138437, 4.4392389],
 'east': [98.6861879, 115.7114992, 106.7800127, 103.7810669, 110.8386897, 106.973975, 123.5519226, 105.0122093, 109.0697907,
  111.8689695, 116.4841801, 114.2054274, 117.900953, 115.8488006, 119.6680859, 118.7650776, 109.6139863, 109.7126785, 
  106.8466516, 135.2282853, 129.7578409, 119.5015499, 125.5500847, 141.0200345, 135.2580601, 103.9803969, 119.8757314,
  122.3438413, 124.9619235, 125.5961913, 127.805486, 101.8861236, 106.6026347, 100.6603006],
 'south': [1.4586943, -8.8504316, -7.4565894, -5.7189866, -8.4159039, -6.3744575, -0.0665628, -2.7822599, -8.0207481,
  -8.4411879, -9.0301357, -4.7153056, -5.4798229, -5.1882715, -2.6537156, 1.1140414, -4.9993635, -1.1825041, -6.4550344,
  -8.5496596, -2.7956171, -9.3098441, -11.2085669, -9.3833333, -4.8517205, -1.1204131, -3.9747153, -7.9722136, -3.3853339,
  -6.9467061, -0.5534119, -3.584461, -4.9241592, -1.0336376],
 'west': [94.7717124, 114.4315327, 104.6513179, 100.330919, 109.901789, 106.3146732, 121.1559965, 101.1225586, 106.0509508,
  108.5558548, 110.8815987, 108.1386521, 113.9911308, 110.6795452, 113.8359298, 114.565164, 104.9996082, 103.065, 103.5068774,
  124.9518369, 123.9232297, 115.5700063, 118.3710228, 133.5217666, 128.4449325, 100.0248488, 117.630978, 116.343821,
  118.8772126, 120.6398645, 123.1123245, 98.2364008, 102.0414013, 96.7203878]}

Province Area and Population Dictionary

In [6]:
province_detail = {
  'area': {0: 57.956, 1: 5.78, 2: 9.663, 3: 19.919, 4: 3.133, 5: 664.0, 6: 11.257, 7: 50.058, 8: 35.378,
  9: 32.801, 10: 47.8, 11: 147.307, 12: 38.744, 13: 153.565, 14: 129.067, 15: 75.468, 16: 16.424, 17: 8.202, 18: 34.624,
  19: 46.914, 20: 31.983, 21: 18.572, 22: 48.718, 23: 319.036, 24: 97.024, 25: 87.024, 26: 16.787, 27: 46.717, 28: 61.841,
  29: 38.068, 30: 13.852, 31: 42.013, 32: 91.592, 33: 72.981},
 'population': {0: '5.316.300', 1: '4.362.000', 2: '12.959.200', 3: '1.971.800', 4: '3.868.600', 5: '10.504.100',
  6: '1.176.400', 7: '3.566.200', 8: '49.023.200', 9: '34.552.500', 10: '39.744.800', 11: '5.045.700', 12: '4.216.300',
  13: '2.649.800', 14: '3.619.700', 15: '695.600', 16: '1.451.100', 17: '2.241.600', 18: '8.457.600', 19: '1.768.500',
  20: '1.235.700', 21: '5.152.400', 22: '5.437.200', 23: '3.347.100', 24: '963.600', 25: '6.835.100', 26: '1.359.200',
  27: '8.819.500', 28: '3.042.100', 29: '2.663.700', 30: '2.512.900', 31: '5.479.500', 32: '8.497.200', 33: '14.639.400'}}

Capital of Province

In [7]:
capital_code = {0: 'Banda Aceh', 1: 'Denpasar', 2: 'Serang', 3: 'Bengkulu', 4: 'Yogyakarta', 5: 'Jakarta', 
                      6: 'Gorontalo', 7: 'Jambi', 8: 'Bandung', 9: 'Semarang', 10: 'Surabaya', 11: 'Pontianak', 
                      12: 'Banjarmasin', 13: 'Palangka Raya', 14: 'Samarinda', 15: 'Tanjung Selor', 
                      16: 'Pangkal Pinang', 17: 'Tanjung Pinang', 18: 'Bandar Lampung', 19: 'Ambon', 20: 'Sofifi', 
                      21: 'Mataram', 22: 'Kupang', 23: 'Jayapura', 24: 'Manokwari', 25: 'Pekanbaru', 
                      26: 'Mamuju', 27: 'Makassar', 28: 'Palu', 29: 'Kendari', 
                      30: 'Manado', 31: 'Padang', 32: 'Palembang', 33 : 'Medan'}

Capital Province Coordinate

In [8]:
capital_coordinate = {'lat': ['5.5528455', '-8.6524973', '-6.10565025', '-3.5186763', '-7.8011945', '-6.1753942', '0.7186174', '-1.611572', '-6.9344694',
                              '-6.9903988', '-7.2459717', '-0.0226903', '-3.3187496', '-2.2072919', '-0.5017804', '2.8469993', '-2.1206733', '0.9236915',
                              '-5.4460713', '47.5542438', '0.734776', '-8.5837726', '-10.1632209', '-2.5387539', '-0.8614456', '0.5262455', '-2.7015755',
                              '-5.1342962', '-0.8957793', '-3.9918068', '1.4900578', '-0.9247587', '-2.9888297', '3.5896654'],
                      'lon': ['95.3192908', '115.2191175', '105.98773335759398', '102.5359834', '110.364917', '106.827183', '122.4555927', '102.779699',
                              '107.6049539', '110.4229104', '112.7378266', '109.3447488', '114.5925828', '113.9164372', '117.1393089', '117.3645691',
                              '106.1134604', '104.446094', '105.2643742', '-2.5566945', '127.5760518', '116.10685', '123.6017755', '140.7037389',
                              '134.0767324', '101.4515727', '118.92039', '119.4124282', '119.8679974', '122.5180066', '124.8408708', '100.3632561',
                              '104.756857', '98.6738261']}

# Province Detail DataFrame

In [ ]:
prov_det_df = pd.read_csv(filepath_or_buffer = 
                              '/content/drive/My Drive/Programming-Project/SAR-COV-19/LSTM-GCN_COVID-19/assets/prov_templates/prov_template_' 
                              + province_code[5] + '.csv')
prov_det_df = prov_det_df.drop(['Unnamed: 0', 'date', 'confirmed', 'confirmed (cumulative)', 'deaths', 'deaths (cumulative)', 'recovered',
                                'recovered (cumulative)'], axis = 1)
index_num = 0
for i in range(34):
  prov_det_dict = {'province': province_code[i],
                    'population': province_detail['population'][i],
                    'area ($km^2$)': province_detail['area'][i],
                    'north border (latitude)': polygon_border['north'][i],
                    'south border (latitude)': polygon_border['south'][i],
                    'east border (longitude)': polygon_border['east'][i],
                    'west border (longitude)': polygon_border['west'][i],
                    'capital': capital_code[i],
                    'capital coordinate (latitude)': capital_coordinate['lat'][i],
                    'capital coordinate (longitude)': capital_coordinate['lon'][i]}
  df = pd.DataFrame(prov_det_dict, index = [index_num])
  index_num += 1
  prov_det_df = pd.concat([prov_det_df, df])

#save dataframe as csv
prov_det_df.to_csv(path_or_buf = '/content/drive/My Drive/Programming-Project/SAR-COV-19/LSTM-GCN_COVID-19/datasets/prov_details_dataframe.csv')

#report
print('prov_details_dataframe.csv saved')

NameError: ignored

# Run to Update

Update Every Province CSV File

In [11]:
#list for logging df length later
prov_len = []

for i in range(34): #iterate through each province in the dictionary
    with urllib.request.urlopen('https://data.covid19.go.id/public/api/prov_detail_' + province_code[i]+ '.json') as url: 
        data = json.loads(url.read().decode())#read json
    
    #coloumns dictionary to store
    columns_dict = {
                    'tanggal': [], 
                    'KASUS': [], 
                    'MENINGGAL': [], 
                    'SEMBUH': []
                   }

    for item in data['list_perkembangan']: #iterate through each date
            for column in columns_dict.keys(): #iterate through each column
                    columns_dict[column].append(item[column]) #add spesific data to the corresponding column
   
    
    #convert dictionary into dataframe
    df = pd.DataFrame(columns_dict)
        
    #date converter
    df['tanggal'] = df['tanggal']/1000 #delete time detail on UNIX Time
    for j in range(len(data['list_perkembangan'])): #iterate through each date
        ts = df['tanggal'][j] 
        ts_conv = datetime.utcfromtimestamp(ts).strftime('%Y-%m-%d %H:%M:%S') #convert to time string
        flipped_ts = ts_conv[8:10] #day
        flipped_ts = flipped_ts + ts_conv[4:8] #month
        flipped_ts = flipped_ts + ts_conv[0:4] #year
        df['tanggal'][j] = flipped_ts
    
    #insert new columns with same value
    df.insert(loc=0, column='province', value=province_code[i])
    df.insert(loc=3, column='confirmed (cumulative)', value=df['KASUS'].cumsum())
    df.insert(loc=5, column='deaths (cumulative)', value=df['MENINGGAL'].cumsum())
    df.insert(loc=7, column='recovered (cumulative)', value=df['SEMBUH'].cumsum())
    df['population'] = province_detail['population'][i]
    df['area'] = province_detail['area'][i]
    df['north'] = polygon_border['north'][i]
    df['south'] = polygon_border['south'][i]
    df['east'] = polygon_border['east'][i]
    df['west'] = polygon_border['west'][i]
    df['capital'] = capital_code[i]
    df['capital_lat'] = capital_coordinate['lat'][i]
    df['capital_lon'] = capital_coordinate['lon'][i]
        
    #rename columns
    df.columns = ['province', 'date', 'confirmed', 'confirmed (cumulative)', 'deaths', 'deaths (cumulative)', 'recovered', 'recovered (cumulative)', 
                  'population', 'area ($km^2$)', 'north border (latitude)','south border (latitude)', 'east border (longitude)',
                  'west border (longitude)', 'capital', 'capital coordinate (latitude)', 'capital coordinate (longitude)']
    
    #read province template csv
    df_template = pd.read_csv(filepath_or_buffer = 
                              '/content/drive/My Drive/Programming-Project/SAR-COV-19/LSTM-GCN_COVID-19/assets/prov_templates/prov_template_' 
                              + province_code[i] + '.csv')
    #clean up
    df_template = df_template.drop(['Unnamed: 0'], axis=1) #drop the unused index column
    
    #concat
    df = pd.concat([df_template, df])

    #clean up 
    df.reset_index(inplace = True) #reset index number
    df = df.drop(['index'], axis=1) #drop the unused index column

    #logging df length for concatenating all province later
    prov_len.append(len(df))

    #save dataframe as csv
    df.to_csv(path_or_buf = '/content/drive/My Drive/Programming-Project/SAR-COV-19/LSTM-GCN_COVID-19/assets/prov_details/prov_detail_' 
                      + province_code[i] + '.csv')
    #report
    print(i, 'prov_detail_' + province_code[i] + '.csv saved')
print('done')

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:32: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:671: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_with_indexer(indexer, value)


0 prov_detail_ACEH.csv saved


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:32: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:671: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_with_indexer(indexer, value)


1 prov_detail_BALI.csv saved
2 prov_detail_BANTEN.csv saved
3 prov_detail_BENGKULU.csv saved
4 prov_detail_DAERAH_ISTIMEWA_YOGYAKARTA.csv saved
5 prov_detail_DKI_JAKARTA.csv saved
6 prov_detail_GORONTALO.csv saved
7 prov_detail_JAMBI.csv saved
8 prov_detail_JAWA_BARAT.csv saved
9 prov_detail_JAWA_TENGAH.csv saved
10 prov_detail_JAWA_TIMUR.csv saved
11 prov_detail_KALIMANTAN_BARAT.csv saved
12 prov_detail_KALIMANTAN_SELATAN.csv saved
13 prov_detail_KALIMANTAN_TENGAH.csv saved
14 prov_detail_KALIMANTAN_TIMUR.csv saved
15 prov_detail_KALIMANTAN_UTARA.csv saved
16 prov_detail_KEPULAUAN_BANGKA_BELITUNG.csv saved
17 prov_detail_KEPULAUAN_RIAU.csv saved
18 prov_detail_LAMPUNG.csv saved
19 prov_detail_MALUKU.csv saved
20 prov_detail_MALUKU_UTARA.csv saved
21 prov_detail_NUSA_TENGGARA_BARAT.csv saved
22 prov_detail_NUSA_TENGGARA_TIMUR.csv saved
23 prov_detail_PAPUA.csv saved
24 prov_detail_PAPUA_BARAT.csv saved
25 prov_detail_RIAU.csv saved
26 prov_detail_SULAWESI_BARAT.csv saved
27 prov_detail

DataFrame Length Check

In [12]:
#print province name that have less length than the maximum length (usually jakarta has the longest dataframe length)
for i in range(33):
  df = pd.read_csv(filepath_or_buffer = 
                              '/content/drive/My Drive/Programming-Project/SAR-COV-19/LSTM-GCN_COVID-19/assets/prov_details/prov_detail_' 
                              + province_code[i] + '.csv')
  if len(df) != max(prov_len):
    print(i, province_code[i], len(df))

0 ACEH 125
3 BENGKULU 127
7 JAMBI 127
11 KALIMANTAN_BARAT 127
15 KALIMANTAN_UTARA 122
17 KEPULAUAN_RIAU 127
20 MALUKU_UTARA 126
22 NUSA_TENGGARA_TIMUR 124
26 SULAWESI_BARAT 126
28 SULAWESI_TENGAH 126
31 SUMATERA_BARAT 127


Concat into National Dataset

In [15]:
#use same length for every province
same_length = True
for a in range(2):
  #assign how many row to use
  if same_length == True:
    df_end = min(prov_len)
  elif same_length == False:
    df_end = -1

  #read first province as base dataframe
  df_national = pd.read_csv(filepath_or_buffer = 
                                '/content/drive/My Drive/Programming-Project/SAR-COV-19/LSTM-GCN_COVID-19/assets/prov_details/prov_detail_' 
                                + province_code[0] + '.csv')
  #cut first province dataframe to desired length
  df_national = df_national[0:df_end]
  #report
  print(1, province_code[0] + ' concatenated')

  #loop throught the rest provinces
  for i in range(1,34):
    #read the next province csv
    df = pd.read_csv(filepath_or_buffer = 
                                '/content/drive/My Drive/Programming-Project/SAR-COV-19/LSTM-GCN_COVID-19/assets/prov_details/prov_detail_' 
                                + province_code[i] + '.csv')
    #concat
    df_national = pd.concat([df_national, df[0:df_end]])
    #report
    print(i, province_code[i] + ' concatenated') 
    #clean up 
    df_national.reset_index(inplace = True) #reset index number
    df_national = df_national.drop(['index'], axis=1) #drop the unused index column

  #set csv file name
  if same_length == True:
    name = 'same_length_main_data'
  elif same_length == False:
    name = 'main_data'
  #save dataframe as csv
  df_national.to_csv(path_or_buf = '/content/drive/My Drive/Programming-Project/SAR-COV-19/LSTM-GCN_COVID-19/datasets/' + name + '.csv')
  #report
  print(name + '.csv saved')
  print()
  same_length = False

1 ACEH concatenated
1 BALI concatenated
2 BANTEN concatenated
3 BENGKULU concatenated
4 DAERAH_ISTIMEWA_YOGYAKARTA concatenated
5 DKI_JAKARTA concatenated
6 GORONTALO concatenated
7 JAMBI concatenated
8 JAWA_BARAT concatenated
9 JAWA_TENGAH concatenated
10 JAWA_TIMUR concatenated
11 KALIMANTAN_BARAT concatenated
12 KALIMANTAN_SELATAN concatenated
13 KALIMANTAN_TENGAH concatenated
14 KALIMANTAN_TIMUR concatenated
15 KALIMANTAN_UTARA concatenated
16 KEPULAUAN_BANGKA_BELITUNG concatenated
17 KEPULAUAN_RIAU concatenated
18 LAMPUNG concatenated
19 MALUKU concatenated
20 MALUKU_UTARA concatenated
21 NUSA_TENGGARA_BARAT concatenated
22 NUSA_TENGGARA_TIMUR concatenated
23 PAPUA concatenated
24 PAPUA_BARAT concatenated
25 RIAU concatenated
26 SULAWESI_BARAT concatenated
27 SULAWESI_SELATAN concatenated
28 SULAWESI_TENGAH concatenated
29 SULAWESI_TENGGARA concatenated
30 SULAWESI_UTARA concatenated
31 SUMATERA_BARAT concatenated
32 SUMATERA_SELATAN concatenated
33 SUMATERA_UTARA concatenated
mai

In [ ]:
df_national

Unnamed: 0  ... capital coordinate (longitude)
0              0  ...                      95.319291
1              1  ...                      95.319291
2              2  ...                      95.319291
3              3  ...                      95.319291
4              4  ...                      95.319291
...          ...  ...                            ...
4143         117  ...                      98.673826
4144         118  ...                      98.673826
4145         119  ...                      98.673826
4146         120  ...                      98.673826
4147         121  ...                      98.673826

[4148 rows x 18 columns]

## Saving Polygon Border Point

In [ ]:
#read map polygon coordinate json file url
with urllib.request.urlopen(
    'https://nominatim.openstreetmap.org/search.php?q=' + province_code[] +' Indonesia&polygon_geojson=1&format=json') as url:
                data = json.loads(url.read().decode())

num_code = 1 #choose which search result to use


polygon_coordinate = {'longitude' : [], 'latitude' : []}
for point in range(len(data[num_code]['geojson']['coordinates'][0])): #iterate through all coordinate point
    polygon_coordinate['longitude'].append(data[num_code]['geojson']['coordinates'][0][point][0]) 
    polygon_coordinate['latitude'].append(data[num_code]['geojson']['coordinates'][0][point][1])

polygon_border['north'].append(max(polygon_coordinate['latitude'])) #northernmost point
polygon_border['east'].append(max(polygon_coordinate['longitude'])) #easternmost point
polygon_border['south'].append(min(polygon_coordinate['latitude'])) #southernmost point
polygon_border['west'].append(min(polygon_coordinate['longitude'])) #westernmost point

prov_code = prov_code + 1 #tells which province next to do

print('north')
print(polygon_border['north'])
print()
print('east')
print(polygon_border['east'])
print()
print('south')
print(polygon_border['south'])
print()
print('west')
print(polygon_border['west'])
print()
print('now province code:', prov_code)

north
[6.2744496, -8.0615889, -5.4996381, -2.2771362, -7.5412887, -4.9993635, 1.3647141, -0.5088816, -4.0387936, -4.0387936, -4.8926893, 2.3146446, -1.3024076, 0.7954077, 2.569023, 4.407823, -0.278773, 4.9918231, -3.7237393, -2.2664176, 3.4464276, -6.7458825, -6.7574264, 1.1369057, 1.2874153, 3.0028667, -0.22745, -1.8906412, 1.583554, -2.7702794, 5.7639518, 0.9073892, -1.5138437, 4.4392389]

east
[98.6861879, 115.7114992, 106.7800127, 103.7810669, 110.8386897, 106.973975, 123.5519226, 105.0122093, 109.0697907, 111.8689695, 116.4841801, 114.2054274, 117.900953, 115.8488006, 119.6680859, 118.7650776, 109.6139863, 109.7126785, 106.8466516, 135.2282853, 129.7578409, 119.5015499, 125.5500847, 141.0200345, 135.2580601, 103.9803969, 119.8757314, 122.3438413, 124.9619235, 125.5961913, 127.805486, 101.8861236, 106.6026347, 100.6603006]

south
[1.4586943, -8.8504316, -7.4565894, -5.7189866, -8.4159039, -6.3744575, -0.0665628, -2.7822599, -8.0207481, -8.4411879, -9.0301357, -4.7153056, -5.4798229

# Base Date Template Maker

In [ ]:
march = [0, 1,  2,  3,  4,  5,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 23, 24, 25, 26, 27, 28, 31, 32, 33]
april = [6, 22, 29, 30]

In [ ]:
for h in range(34): #iterate through each province in the dictionary
    i = 5
    with urllib.request.urlopen('https://data.covid19.go.id/public/api/prov_detail_' + province_code[i]+ '.json') as url: 
        data = json.loads(url.read().decode())#read json
    print('loaded '+province_code[i], i)
    #coloumns dictionary to store
    columns_dict = {
                    'tanggal': [], 
                    'KASUS': [], 
                    'MENINGGAL': [], 
                    'SEMBUH': []
                   }
    
    #convert dictionary into dataframe
    df = pd.DataFrame(columns_dict)
        
    ts = data['list_perkembangan'][0]['tanggal']/1000
    ts_conv = datetime.utcfromtimestamp(ts).strftime('%Y-%m-%d %H:%M:%S') #convert to time string
    flipped_ts = ts_conv[8:10] #day
    flipped_ts = flipped_ts + ts_conv[4:8] #month
    flipped_ts = flipped_ts + ts_conv[0:4] #year
    case_zero = flipped_ts
    print(case_zero)
    #start date format
    dd = int(case_zero[0:2])
    
    #for province that start from april
    if case_zero[4] == '4':
      for j in range(int(case_zero[0:2])-1): #loop for april
          dd -= 1
          minus_one = str(dd) + '-04-2020'
          if len(minus_one) == 9:
              minus_one = '0' + minus_one
          template_dict = {
                          'tanggal': minus_one, #date
                          'KASUS': 0, #confirmed
                          'MENINGGAL': 0, #deaths
                          'SEMBUH': 0 #recovered
                          }
          template_df = pd.DataFrame(template_dict, index = [0]) #make a dataframe with one row of minus_one
          df = pd.concat([template_df, df]) #concat

      dd = 32
      for k in range(31): #loop for whole march
          dd -= 1
          minus_one = str(dd) + '-03-2020'
          if len(minus_one) == 9:
              minus_one = '0' + minus_one
          template_dict = {
                          'tanggal': minus_one, #date
                          'KASUS': 0, #confirmed
                          'MENINGGAL': 0, #deaths
                          'SEMBUH': 0 #recovered
                          }
          template_df = pd.DataFrame(template_dict, index = [0]) #make a dataframe with one row of minus_one
          df = pd.concat([template_df, df]) #concat
        
    #for province that start from april
    elif case_zero[4] == '3':
      for l in range(int(case_zero[0:2])-1): #loop for march
          dd -= 1
          minus_one = str(dd) + '-03-2020'
          if len(minus_one) == 9:
              minus_one = '0' + minus_one
          template_dict = {
                          'tanggal': minus_one, #date
                          'KASUS': 0, #confirmed
                          'MENINGGAL': 0, #deaths
                          'SEMBUH': 0 #recovered
                          }
          template_df = pd.DataFrame(template_dict, index = [0]) #make a dataframe with one row of minus_one
          df = pd.concat([template_df, df]) #concat
    
    #insert new column with same value, the province name
    df.insert(loc=0, column='province', value=province_code[i])
    df.insert(loc=3, column='confirmed (cumulative)', value=df['KASUS'].cumsum())
    df.insert(loc=5, column='deaths (cumulative)', value=df['MENINGGAL'].cumsum())
    df.insert(loc=7, column='recovered (cumulative)', value=df['SEMBUH'].cumsum())
    df['population'] = province_detail['population'][i]
    df['area'] = province_detail['area'][i]
    df['north'] = polygon_border['north'][i]
    df['south'] = polygon_border['south'][i]
    df['east'] = polygon_border['east'][i]
    df['west'] = polygon_border['west'][i]
    df['capital'] = capital_code[i]
    df['capital_lat'] = capital_coordinate['lat'][i]
    df['capital_lon'] = capital_coordinate['lon'][i]

    #rename the columns
    df.columns = ['province', 'date', 'confirmed', 'confirmed (cumulative)', 'deaths', 'deaths (cumulative)', 'recovered', 'recovered (cumulative)', 
                  'population', 'area ($km^2$)', 'north border (latitude)','south border (latitude)', 'east border (longitude)',
                  'west border (longitude)', 'capital', 'capital coordinate (latitude)', 'capital coordinate (longitude)']
    

    #clean up 
    df.reset_index(inplace = True) #reset index number
    df = df.drop(['index'], axis=1) #drop the unused index column

    #save dataframe as csv
    df.to_csv(path_or_buf = 
                       '/content/drive/My Drive/Programming-Project/SAR-COV-19/LSTM-GCN_COVID-19/assets/prov_templates/prov_template_' 
                       + province_code[i] + '.csv')
    #report
    print('prov_template_' + province_code[i] + '.csv saved')
    print()

In [ ]:
#check
def tester(df, i):
    if df['date'][0] != '01-03-2020':
      print(i, province_code[i] + ' ' + df['date'][0])

for i in range(34):
    df_test = pd.read_csv(filepath_or_buffer = 
                          '/content/drive/My Drive/Programming-Project/SAR-COV-19/LSTM-GCN_COVID-19/assets/prov_templates/prov_template_' 
                          + province_code[i] + '.csv')
    try:
      tester(df_test, i)
    except IndexError: #except for month that starts from March 1st
      print(i, province_code[i])
      continue

#Saving Province Capital Coordinate

In [ ]:
for i in range(34):
  capital_name = capital_code[i].replace(' ', '+')
  with urllib.request.urlopen(
      'https://nominatim.openstreetmap.org/search.php?q=' + capital_name +'&polygon_geojson=1&format=json') as url:
                  data = json.loads(url.read().decode())

  capital_coordinate['lat'].append(data[0]['lat'])
  capital_coordinate['lon'].append(data[0]['lon'])

# Adjacency Matrix

In [ ]:
#adjacency dictionary
adj_dict = {'ACEH': [
                     'SUMATERA_UTARA'
                    ],
            
            'BALI': [
                     'JAWA_TIMUR',
                     'NUSA_TENGGARA_BARAT'
                    ],
            
            'BANTEN': [
                       'LAMPUNG',
                       'DKI_JAKARTA',
                       'JAWA_BARAT'
                      ],
            
            'BENGKULU': [
                         'SUMATERA_BARAT',
                         'JAMBI',
                         'SUMATERA_SELATAN',
                         'LAMPUNG'
                        ],
            
            'DAERAH_ISTIMEWA_YOGYAKARTA': [
                                           'JAWA_TENGAH'
                                          ],
            
            'DKI_JAKARTA': [
                            'LAMPUNG',
                            'BANTEN',
                            'JAWA_BARAT'
                           ],
            
            'GORONTALO': [
                          'SULAWESI_UTARA',
                          'SULAWESI_TENGAH'
                         ],
            
            'JAMBI': [
                      'RIAU',
                      'KEPULAUAN_RIAU',
                      'KEPULAUAN_BANGKA_BELITUNG',
                      'SUMATERA_SELATAN',
                      'BENGKULU',
                      'SUMATERA_BARAT'
                     ],
            
            'JAWA_BARAT': [
                           'BANTEN',
                           'DKI_JAKARTA',
                           'KEPULAUAN_BANGKA_BELITUNG',
                           'JAWA_TENGAH'
                          ],
            
            'JAWA_TENGAH': [
                            'JAWA_BARAT',
                            'KALIMANTAN_BARAT',
                            'KALIMANTAN_TENGAH',
                            'JAWA_TIMUR',
                            'DAERAH_ISTIMEWA_YOGYAKARTA'
                           ],
            
            'JAWA_TIMUR': [
                           'JAWA_TENGAH',
                           'KALIMANTAN_TENGAH',
                           'KALIMANTAN_SELATAN',
                           'SULAWESI_SELATAN',
                           'BALI',
                           'NUSA_TENGGARA_BARAT'
                          ],
            
            'KALIMANTAN_BARAT': [
                                 'KEPULAUAN_RIAU',
                                 'KEPULAUAN_BANGKA_BELITUNG',
                                 'JAWA_TENGAH',
                                 'KALIMANTAN_TENGAH',
                                 'KALIMANTAN_TIMUR'
                                ],
            
            'KALIMANTAN_SELATAN': [
                                   'KALIMANTAN_TIMUR',
                                   'SULAWESI_BARAT',
                                   'JAWA_TIMUR',
                                   'SULAWESI_SELATAN',
                                   'JAWA_TIMUR',
                                   'KALIMANTAN_TENGAH'
                                  ],
            
            'KALIMANTAN_TENGAH': [
                                  'KALIMANTAN_TIMUR',
                                  'KALIMANTAN_SELATAN',
                                  'JAWA_TIMUR',
                                  'JAWA_TENGAH',
                                  'KALIMANTAN_BARAT'
                                 ],
            
            'KALIMANTAN_TIMUR': [
                                 'KALIMANTAN_UTARA',
                                 'KALIMANTAN_BARAT',
                                 'KALIMANTAN_TENGAH',
                                 'KALIMANTAN_SELATAN',
                                 'SULAWESI_BARAT',
                                 'SULAWESI_TENGAH'
                                ],
            
            'KALIMANTAN_UTARA': [
                                 'KALIMANTAN_TIMUR'
                                ],
            
            'KEPULAUAN_BANGKA_BELITUNG': [
                                          'KEPULAUAN_RIAU',
                                          'KALIMANTAN_BARAT',
                                          'JAWA_BARAT',
                                          'LAMPUNG',
                                          'SUMATERA_SELATAN',
                                          'JAMBI'
                                         ],
            
            'KEPULAUAN_RIAU': [
                               'RIAU',
                               'JAMBI',
                               'KEPULAUAN_BANGKA_BELITUNG',
                               'KALIMANTAN_BARAT'
                              ],
            
            'LAMPUNG': [
                        'BENGKULU',
                        'SUMATERA_SELATAN',
                        'KEPULAUAN_BANGKA_BELITUNG',
                        'DKI_JAKARTA',
                        'BANTEN'
                       ],
            
            'MALUKU': [
                       'NUSA_TENGGARA_TIMUR',
                       'SULAWESI_TENGGARA',
                       'SULAWESI_TENGAH',
                       'MALUKU_UTARA',
                       'PAPUA_BARAT',
                       'PAPUA'
                      ],
            
            'MALUKU_UTARA': [
                             'SULAWESI_UTARA',
                             'SULAWESI_TENGAH',
                             'MALUKU',
                             'PAPUA_BARAT'
                            ],
            
            'NUSA_TENGGARA_BARAT': [
                                    'BALI',
                                    'JAWA_TIMUR',
                                    'SULAWESI_SELATAN',
                                    'NUSA_TENGGARA_TIMUR'
                                   ],
            
            'NUSA_TENGGARA_TIMUR': [
                                    'NUSA_TENGGARA_BARAT',
                                    'SULAWESI_SELATAN',
                                    'SULAWESI_TENGGARA',
                                    'MALUKU'
                                   ],
            
            'PAPUA': [
                      'PAPUA_BARAT',
                      'MALUKU'
                     ],
            
            'PAPUA_BARAT': [
                            'PAPUA',
                            'MALUKU',
                            'MALUKU_UTARA'
                           ],
            
            'RIAU': [
                     'SUMATERA_UTARA',
                     'SUMATERA_BARAT',
                     'JAMBI',
                     'KEPULAUAN_RIAU'
                    ],
            
            'SULAWESI_BARAT': [
                               'KALIMANTAN_TIMUR',
                               'SULAWESI_TENGAH',
                               'SULAWESI_SELATAN',
                               'KALIMANTAN_TIMUR'
                              ],
            
            'SULAWESI_SELATAN': [
                                 'SULAWESI_BARAT',
                                 'SULAWESI_TENGAH',
                                 'SULAWESI_TENGGARA',
                                 'NUSA_TENGGARA_TIMUR',
                                 'NUSA_TENGGARA_BARAT',
                                 'JAWA_TIMUR',
                                 'KALIMANTAN_SELATAN'
                                ],
            
            'SULAWESI_TENGAH': [
                                'GORONTALO',
                                'SULAWESI_UTARA',
                                'MALUKU_UTARA',
                                'MALUKU',
                                'SULAWESI_TENGGARA',
                                'SULAWESI_SELATAN',
                                'SULAWESI_BARAT',
                                'KALIMANTAN_TIMUR'
                               ],
            
            'SULAWESI_TENGGARA': [
                                  'SULAWESI_TENGAH',
                                  'MALUKU',
                                  'NUSA_TENGGARA_TIMUR',
                                  'SULAWESI_SELATAN'
                                 ],
            
            'SULAWESI_UTARA': [
                               'GORONTALO',
                               'SULAWESI_TENGAH',
                               'MALUKU_UTARA'
                              ],
            
            'SUMATERA_BARAT': [
                               'SUMATERA_UTARA',
                               'RIAU',
                               'JAMBI',
                               'BENGKULU'
                              ],
            
            'SUMATERA_SELATAN': [
                                 'BENGKULU',
                                 'JAMBI',
                                 'KEPULAUAN_BANGKA_BELITUNG',
                                 'LAMPUNG'
                                ],
            
            'SUMATERA_UTARA': [
                               'ACEH',
                               'SUMATERA_BARAT',
                               'RIAU'
                              ]
    }

In [ ]:
#function to translate value to key in province_code
def get_key(val): 
    for key, value in province_code.items(): 
         if val == value: 
             return key 
  
    return "key doesn't exist" + val

In [ ]:
#translate adjacency dictionary to number list
province_name = adj_dict.keys()
for i in province_name:
    for j in adj_dict[i]:
        src.append(get_key(i))
        dst.append(get_key(j))

In [ ]:
src = [0, 1, 1, 2, 2, 2, 3, 3, 3, 3, 4, 5, 5, 5, 6, 6, 7, 7, 7, 7, 7, 7, 8, 8, 8, 8, 9, 9, 9, 9, 9, 10, 10, 10, 10, 10, 
       10, 11, 11, 11, 11, 11, 12, 12, 12, 12, 12, 12, 13, 13, 13, 13, 13, 14, 14, 14, 14, 14, 14, 15, 16, 16, 16, 16, 
       16, 16, 17, 17, 17, 17, 18, 18, 18, 18, 18, 19, 19, 19, 19, 19, 19, 20, 20, 20, 20, 21, 21, 21, 21, 22, 22, 22, 
       22, 23, 23, 24, 24, 24, 25, 25, 25, 25, 26, 26, 26, 26, 27, 27, 27, 27, 27, 27, 27, 28, 28, 28, 28, 28, 28, 28, 
       28, 29, 29, 29, 29, 30, 30, 30, 31, 31, 31, 31, 32, 32, 32, 32, 33, 33, 33]

In [ ]:
dst = [33, 10, 21, 18, 5, 8, 31, 7, 32, 18, 9, 18, 2, 8, 30, 28, 25, 17, 16, 32, 3, 31, 2, 5, 16, 9, 8, 11, 13, 10, 4, 9,
       13, 12, 27, 1, 21, 17, 16, 9, 13, 14, 14, 26, 10, 27, 10, 13, 14, 12, 10, 9, 11, 15, 11, 13, 12, 26, 28, 14, 17, 
       11, 8, 18, 32, 7, 25, 7, 16, 11, 3, 32, 16, 5, 2, 22, 29, 28, 20, 24, 23, 30, 28, 19, 24, 1, 10, 27, 22, 21, 27, 
       29, 19, 24, 19, 23, 19, 20, 33, 31, 7, 17, 14, 28, 27, 14, 26, 28, 29, 22, 21, 10, 12, 6, 30, 20, 19, 29, 27, 26, 
       14, 28, 19, 22, 27, 6, 28, 20, 33, 25, 7, 3, 3, 7, 16, 18, 0, 31, 25]

In [ ]:
#make a dataframe with province names as column and row
adj_df = pd.DataFrame(province_code, index = province_code.values())
adj_df.columns = province_code.values()
adj_df

ACEH  BALI  BANTEN  BENGKULU  \
ACEH                        ACEH  BALI  BANTEN  BENGKULU   
BALI                        ACEH  BALI  BANTEN  BENGKULU   
BANTEN                      ACEH  BALI  BANTEN  BENGKULU   
BENGKULU                    ACEH  BALI  BANTEN  BENGKULU   
DAERAH_ISTIMEWA_YOGYAKARTA  ACEH  BALI  BANTEN  BENGKULU   
DKI_JAKARTA                 ACEH  BALI  BANTEN  BENGKULU   
GORONTALO                   ACEH  BALI  BANTEN  BENGKULU   
JAMBI                       ACEH  BALI  BANTEN  BENGKULU   
JAWA_BARAT                  ACEH  BALI  BANTEN  BENGKULU   
JAWA_TENGAH                 ACEH  BALI  BANTEN  BENGKULU   
JAWA_TIMUR                  ACEH  BALI  BANTEN  BENGKULU   
KALIMANTAN_BARAT            ACEH  BALI  BANTEN  BENGKULU   
KALIMANTAN_SELATAN          ACEH  BALI  BANTEN  BENGKULU   
KALIMANTAN_TENGAH           ACEH  BALI  BANTEN  BENGKULU   
KALIMANTAN_TIMUR            ACEH  BALI  BANTEN  BENGKULU   
KALIMANTAN_UTARA            ACEH  BALI  BANTEN  BENGKULU   
KEPULAUAN_BANGKA_BELITUNG   ACEH  BALI  BANTEN  BENGKULU   
KEPULAUAN_RIAU              ACEH  BALI  BANTEN  BENGKULU   
LAMPUNG                     ACEH  BALI  BANTEN  BENGKULU   
MALUKU                      ACEH  BALI  BANTEN  BENGKULU   
MALUKU_UTARA                ACEH  BALI  BANTEN  BENGKULU   
NUSA_TENGGARA_BARAT         ACEH  BALI  BANTEN  BENGKULU   
NUSA_TENGGARA_TIMUR         ACEH  BALI  BANTEN  BENGKULU   
PAPUA                       ACEH  BALI  BANTEN  BENGKULU   
PAPUA_BARAT                 ACEH  BALI  BANTEN  BENGKULU   
RIAU                        ACEH  BALI  BANTEN  BENGKULU   
SULAWESI_BARAT              ACEH  BALI  BANTEN  BENGKULU   
SULAWESI_SELATAN            ACEH  BALI  BANTEN  BENGKULU   
SULAWESI_TENGAH             ACEH  BALI  BANTEN  BENGKULU   
SULAWESI_TENGGARA           ACEH  BALI  BANTEN  BENGKULU   
SULAWESI_UTARA              ACEH  BALI  BANTEN  BENGKULU   
SUMATERA_BARAT              ACEH  BALI  BANTEN  BENGKULU   
SUMATERA_SELATAN            ACEH  BALI  BANTEN  BENGKULU   
SUMATERA_UTARA              ACEH  BALI  BANTEN  BENGKULU   

                            DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA  \
ACEH                        DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
BALI                        DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
BANTEN                      DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
BENGKULU                    DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
DAERAH_ISTIMEWA_YOGYAKARTA  DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
DKI_JAKARTA                 DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
GORONTALO                   DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
JAMBI                       DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
JAWA_BARAT                  DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
JAWA_TENGAH                 DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
JAWA_TIMUR                  DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
KALIMANTAN_BARAT            DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
KALIMANTAN_SELATAN          DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
KALIMANTAN_TENGAH           DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
KALIMANTAN_TIMUR            DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
KALIMANTAN_UTARA            DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
KEPULAUAN_BANGKA_BELITUNG   DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
KEPULAUAN_RIAU              DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
LAMPUNG                     DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
MALUKU                      DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
MALUKU_UTARA                DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
NUSA_TENGGARA_BARAT         DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
NUSA_TENGGARA_TIMUR         DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
PAPUA                       DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
PAPUA_BARAT                 DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
RIAU                        DAERAH_ISTIMEWA_YOGYAKARTA  DKI_JAKARTA   
SULAWESI_B

In [ ]:
#change all value to 0
for col in province_code.values():
    for row in province_code.values():
        adj_df[col][row] = 0
        
#change all value that is adjacent to each other to 1
for i in range(len(src)):
        adj_df[province_code[src[i]]][province_code[dst[i]]] = 1

In [ ]:
adj_df

ACEH BALI BANTEN BENGKULU  \
ACEH                          0    0      0        0   
BALI                          0    0      0        0   
BANTEN                        0    0      0        0   
BENGKULU                      0    0      0        0   
DAERAH_ISTIMEWA_YOGYAKARTA    0    0      0        0   
DKI_JAKARTA                   0    0      1        0   
GORONTALO                     0    0      0        0   
JAMBI                         0    0      0        1   
JAWA_BARAT                    0    0      1        0   
JAWA_TENGAH                   0    0      0        0   
JAWA_TIMUR                    0    1      0        0   
KALIMANTAN_BARAT              0    0      0        0   
KALIMANTAN_SELATAN            0    0      0        0   
KALIMANTAN_TENGAH             0    0      0        0   
KALIMANTAN_TIMUR              0    0      0        0   
KALIMANTAN_UTARA              0    0      0        0   
KEPULAUAN_BANGKA_BELITUNG     0    0      0        0   
KEPULAUAN_RIAU                0    0      0        0   
LAMPUNG                       0    0      1        1   
MALUKU                        0    0      0        0   
MALUKU_UTARA                  0    0      0        0   
NUSA_TENGGARA_BARAT           0    1      0        0   
NUSA_TENGGARA_TIMUR           0    0      0        0   
PAPUA                         0    0      0        0   
PAPUA_BARAT                   0    0      0        0   
RIAU                          0    0      0        0   
SULAWESI_BARAT                0    0      0        0   
SULAWESI_SELATAN              0    0      0        0   
SULAWESI_TENGAH               0    0      0        0   
SULAWESI_TENGGARA             0    0      0        0   
SULAWESI_UTARA                0    0      0        0   
SUMATERA_BARAT                0    0      0        1   
SUMATERA_SELATAN              0    0      0        1   
SUMATERA_UTARA                1    0      0        0   

                           DAERAH_ISTIMEWA_YOGYAKARTA DKI_JAKARTA GORONTALO  \
ACEH                                                0           0         0   
BALI                                                0           0         0   
BANTEN                                              0           1         0   
BENGKULU                                            0           0         0   
DAERAH_ISTIMEWA_YOGYAKARTA                          0           0         0   
DKI_JAKARTA                                         0           0         0   
GORONTALO                                           0           0         0   
JAMBI                                               0           0         0   
JAWA_BARAT                                          0           1         0   
JAWA_TENGAH                                         1           0         0   
JAWA_TIMUR                                          0           0         0   
KALIMANTAN_BARAT                                    0           0         0   
KALIMANTAN_SELATAN                                  0           0         0   
KALIMANTAN_TENGAH                                   0           0         0   
KALIMANTAN_TIMUR                                    0           0         0   
KALIMANTAN_UTARA                                    0           0         0   
KEPULAUAN_BANGKA_BELITUNG                           0           0         0   
KEPULAUAN_RIAU                                      0           0         0   
LAMPUNG                                             0           1         0   
MALUKU                                              0           0         0   
MALUKU_UTARA                                        0           0         0   
NUSA_TENGGARA_BARAT                                 0           0         0   
NUSA_TENGGARA_TIMUR                                 0           0         0   
PAPUA                                               0           0         0   
PAPUA_BARAT                                         0           0         0   
RIAU        

In [ ]:
adj_df.to_csv(path_or_buf = 'E:\\Tugas\\SAR COV 19 in Indonesia dataset\\adjacency_matrix.csv') #save dataframe to csv
print('adjacency_matrix.csv saved') #report

# Update

Terakhir:

udah berhasil baca json tiap provinsi,

ubah ke dataframe,

dan save sebagai csv,

bisa ubah UNIX Time ke readable string,

province yang ga ada tanggal pelaporan ga ketauan dari json

error di date converter(cuma warning)

download json koordinat tiap provinsi, download versi yang sama

adjacency matrix

nama, koordinat ibu kota provinsi

template base tanggal (maret)

kumulatif

template base tanggal (april)

potong df dengan panjang yang sama sesuai panjang minimum

-


Selanjutnya:

detail prov jabar




## DataFrame Error
<ipython-input-86-110297c19cf4>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['tanggal'][i] = ts_conv[0:10]
c:\users\user\appdata\local\programs\python\python38-32\lib\site-packages\pandas\core\indexing.py:671: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_with_indexer(indexer, value)